beamer/config/commands.py

```diff
@@ -131,3 +138,255 @@ def read(
     config.block = fetcher.synced_block
     config.to_file(state_path)
     log.info("Stored configuration", path=str(state_path))
+
+
+def _generate_fee_updates(
+    current_config: Configuration, desired_config: Configuration
+) -> Generator[tuple, None, None]:
+    current_rm = current_config.request_manager
+    desired_rm = desired_config.request_manager
+
+    # Fees. Note the order of elements in the tuple must match the order of contract function
+    # arguments.
+    current_fees = current_rm.min_fee_ppm, current_rm.lp_fee_ppm, current_rm.protocol_fee_ppm
+    desired_fees = desired_rm.min_fee_ppm, desired_rm.lp_fee_ppm, desired_rm.protocol_fee_ppm
+    if current_fees != desired_fees:
+        yield "RequestManager", "updateFees", *desired_fees
+
+
+def _generate_chain_updates(
+    current_config: Configuration, desired_config: Configuration
+) -> Generator[tuple, None, None]:
+    current_chains = current_config.request_manager.chains
+    desired_chains = desired_config.request_manager.chains
+
+    # We don't support chain removals yet.
+    assert set(current_chains).issubset(desired_chains)
+
+    for chain_id, desired_chain in desired_chains.items():
+        current_chain = current_chains.get(chain_id)
+        if current_chain is None or current_chain != desired_chain:
+            yield (
+                "RequestManager",
+                "updateChain",
+                chain_id,
+                desired_chain.finality_period,
+                desired_chain.transfer_cost,
+                desired_chain.target_weight_ppm,
+            )
+
+
+def _generate_token_updates(
+    current_config: Configuration, desired_config: Configuration
+) -> Generator[tuple, None, None]:
+    current_tokens = current_config.request_manager.tokens
+    desired_tokens = desired_config.request_manager.tokens
+
+    # First check for removed tokens.
+    for symbol in current_tokens:
+        if symbol not in desired_tokens:
+            address = current_config.token_addresses[symbol]
+            # The token is disabled by setting its transfer limit to zero,
+            # but we set also eth_in_token to zero.
+            yield "RequestManager", "updateToken", address, 0, 0
+
+    # Then check for newly added and modified tokens.
+    for symbol, desired_token in desired_tokens.items():
+        current_token = current_tokens.get(symbol)
+        if current_token is None or current_token != desired_token:
+            address = desired_config.token_addresses[symbol]
```

Comment:
Can't we do this check on validation of the config? I would consider a config where the token mapping from symbol to addresses for a given token is missing, an invalid config

Reply:
> Can't we do this check on validation of the config? I would consider a config where the token mapping from symbol to addresses for a given token is missing, an invalid config

We already have that check when loading the configuration.

This is about something else: when we generate updates for tokens, we cannot assume that a desired token e.g. `TST` already exists in the current configuration. So we need to check first whether `current_token is None` -- if it is, that means we have a new token in the desired configuration, for which we need to issue a transaction.

Similarly, if `current_token != desired_token` (note: `current_token` and `desired_token` are of type `TokenConfig`) then the token does exist in the current configuration, but its configuration differs from the desired one, which means we need to issue a transaction.

These two cases are what the comment above the `for` loop points at:
```
# Then check for newly added and modified tokens.
```

I hope that clarifies things.